src/phpDocumentor/FilesystemFactory.php

```diff
@@ -0,0 +1,64 @@
+<?php
+/**
+ * This file is part of phpDocumentor.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @copyright 2010-2015 Mike van Riel<mike@phpdoc.org>
+ * @license   http://www.opensource.org/licenses/mit-license.php MIT
+ * @link      http://phpdoc.org
+ */
+
+namespace phpDocumentor;
+
+use \LogicException;
+use League\Flysystem\Adapter\Local;
+use League\Flysystem\Filesystem;
+use League\Flysystem\MountManager;
+
+/**
+ * Class FilesystemFactory
+ */
+final class FilesystemFactory
+{
+    /** @var MountManager */
+    private $mountManager;
+
+    /**
+     * @param MountManager $mountManager
+     */
+    public function __construct(MountManager $mountManager)
+    {
+        $this->mountManager = $mountManager;
+    }
+
+    /**
+     * Returns a Filesystem instance based on the scheme of the provided Dsn
+     *
+     * @param Dsn $dsn
+     * @return Filesystem
+     */
+    public function create(Dsn $dsn)
+    {
+        $dsnId = hash('md5', (string)$dsn);
+
+        try {
+            $filesystem = $this->mountManager->getFilesystem($dsnId);
+        } catch (LogicException $e) {
```

Comment:
Instead of relying on exceptions (which are circumstantial evidence because you do not know for sure that a LogicException means the FS is missing) you can use the `has()` method of the mount manager to check if your intended FS already exists.

By using `has` you can rely on exceptions to be exceptions in the logic and not part of your business logic.